Pin single docker base image from commandline argument

oh6hay opened the pull request:
Fetch the latest hash for a single docker base image supplied as commandline argument.

Example:
```
$ dockpin docker-base ubuntu:16.04
ubuntu@sha256:0f71fa8d4d2d4292c3c617fda2b36f6dabe5c8b6e34c3dc5b0d17d4e704bd39c
```

This makes it easier to pin docker base images in other files than Dockerfiles, eg. circleci yml files, [like this](https://gist.github.com/oh6hay/a7195da72c7de5a389cfa280a97ff07d)

oh6hay committed:
Pin single docker base image from commandline argument

Fetch the latest hash for a single docker base image supplied as
commandline argument.

Example:
```
$ dockpin docker-base ubuntu:16.04
ubuntu@sha256:0f71fa8d4d2d4292c3c617fda2b36f6dabe5c8b6e34c3dc5b0d17d4e704bd39c
```

Jille commented:
Interesting feature idea. How about making the command "dockpin docker resolve <image>"?

I'd love to just add regular support for updating circleci files so that people don't have to write their own shell script for it. WDYT?

Also, should the output be ubuntu@... or ubuntu:16.04@... ? It might be nice for future readers to know what tag it's based on?

oh6hay commented:
hi @Jille !

`dockpin docker resolve` sounds good -- I'll change that.

Updating circleci files directly sounds good. I'm not sure how to do that but i'll look into it for the learning experience. (Still learning go)

Commit:
Small fixes

- renamed the `docker-base` command, it's now `resolve`
- added resolve command to the docker command, not root command

Jille commented on docker.go:
```diff
@@ -26,6 +27,15 @@ var (
 	}
 
+	dockerBaseCmd = &cobra.Command{
+		Use:          "resolve [base image]",
```
Let's make this `[base-image]` so people/Cobra understand this is a single argument.

Jille commented on docker.go:
```diff
@@ -57,6 +67,21 @@ func runDockerPin(cmd *cobra.Command, args []string) error {
+	if diErr != nil {
+		return diErr
+	}
+	hashedBase := baseImage + "@" + string(di.Descriptor.Digest) + "\n"
```
Should we leave the tag included here? I suppose it's nice for readers to see what the image is based on, so they know how large the upgrade they're gonna do is.

oh6hay replied:
good point, of course the tag should stay

Jille commented on docker.go:
```diff
@@ -57,6 +67,21 @@ func runDockerPin(cmd *cobra.Command, args []string) error {
+	if err != nil {
+		return err
+	}
+	di, diErr := dockerClient.DistributionInspect(cmd.Context(), baseImageAndVersion, "")
```
You can redefine variables in Go (with the same type), just call `diErr` `err` and that'll work.

Jille commented on docker.go:
```diff
@@ -26,6 +27,15 @@ var (
 		RunE:         runDockerPin,
 	}
 
+	dockerBaseCmd = &cobra.Command{
```
Let's call this `dockerResolveCmd`

Jille commented on docker.go:
```diff
@@ -26,6 +27,15 @@ var (
+		Short:        "Prints the current digest of the given base image",
+		Args:         cobra.ExactArgs(1),
+		SilenceUsage: false,
+		RunE:         runDockerBasePin,
```
Let's call this `runDockerResolve`

Jille commented on docker.go:
```diff
@@ -57,6 +67,21 @@ func runDockerPin(cmd *cobra.Command, args []string) error {
+		return diErr
+	}
+	hashedBase := baseImage + "@" + string(di.Descriptor.Digest) + "\n"
+	return ioutil.WriteFile("/dev/stdout", []byte(hashedBase), 0644)
```
Just use `fmt.Println(hashedBase)` instead of opening+writing+closing /dev/stdout.

oh6hay replied:
oops

Jille commented on docker.go:
```diff
@@ -26,6 +27,15 @@ var (
+		Example:      "resolve ubuntu:20.04",
+		Short:        "Prints the current digest of the given base image",
+		Args:         cobra.ExactArgs(1),
+		SilenceUsage: false,
```
SilenceUsage should probably be true, because we don't want to print the usage in case we return an error from the implementation.

oh6hay committed:
Implement suggested improvements

oh6hay commented:
Thanks for the feedback!

oh6hay committed:
Implement suggested improvements

oh6hay committed:
Println

Jille commented on docker.go:
```diff
@@ -57,6 +66,21 @@ func runDockerPin(cmd *cobra.Command, args []string) error {
+		return err
+	}
+	hashedBase := baseImageAndVersion + "@" + string(di.Descriptor.Digest)
+	_, err = fmt.Println(hashedBase)
```
Congratulations, you just became the first person ever to check fmt.Println's return value ;)

Jille commented:
I had to manually merge it to prevent me from becoming the author of your code :/
